src/Database/PostgreSQL/PQTypes/Model/Trigger.hs

```diff
@@ -0,0 +1,301 @@
+-- the attached function, the entire query that had created the trigger is received using
+-- @pg_get_triggerdef(t.oid, true)::text@ and then parsed. The result of that call will be
+-- decompiled and normalized, which means that it's likely not what the user had
+-- originally typed.
+--
+-- @since 1.15.0
+getDBTriggers :: forall m. MonadDB m => RawSQL () -> m [Trigger]
+getDBTriggers tableName = do
+  runQuery_ . sqlSelect "pg_trigger t" $ do
+    sqlResult "t.tgname::text" -- name
+    sqlResult "t.tgtype" -- smallint == int2 => (2 bytes)
+    sqlResult "t.tgdeferrable" -- boolean
+    sqlResult "t.tginitdeferred"-- boolean
+    -- This gets the entire query that created this trigger. Note that it's decompiled and
+    -- normalized, which means that it's likely not what the user actually typed. For
+    -- example, if the original query had excessive whitespace in it, it won't be in this
+    -- result.
+    sqlResult "pg_get_triggerdef(t.oid, true)::text"
+    sqlResult "p.proname::text" -- name
+    sqlResult "p.prosrc" -- text
+    sqlResult "c.relname::text"
+    sqlJoinOn "pg_proc p" "t.tgfoid = p.oid"
+    sqlJoinOn "pg_class c" "c.oid = t.tgrelid"
+    sqlWhereEq "t.tgisinternal" False
+    sqlWhereEq "c.relname" $ unRawSQL tableName
+  fetchMany getTrigger
+  where
+    getTrigger :: (String, Int16, Bool, Bool, String, String, String, String) -> Trigger
+    getTrigger (tgname, tgtype, tgdeferrable, tginitdeferrable, triggerdef, proname, prosrc, tblName) =
+      Trigger { triggerTable = tableName'
+              , triggerName = triggerBaseName (unsafeSQL tgname) tableName'
+              , triggerEvents = trgEvents
+              , triggerDeferrable = tgdeferrable
+              , triggerInitiallyDeferred = tginitdeferrable
+              , triggerWhen = tgrWhen
+              , triggerFunction = TriggerFunction (unsafeSQL proname) (unsafeSQL prosrc)
+              }
+      where
+        tableName' :: RawSQL ()
+        tableName' = unsafeSQL tblName
+
+        parseBetween :: Text -> Text -> Maybe (RawSQL ())
+        parseBetween left right =
+          let (prefix, match) = Text.breakOnEnd left $ Text.pack triggerdef
+          in if Text.null prefix
+             then Nothing
+             else Just $ (rawSQL . fst $ Text.breakOn right match) ()
+
+        -- Get the WHEN part of the query. Anything between WHEN and EXECUTE is what we
+        -- want. The Postgres' grammar guarantees that WHEN and EXECUTE are always next to
+        -- each other and in that order.
+        tgrWhen :: Maybe (RawSQL ())
+        tgrWhen = parseBetween "WHEN (" ") EXECUTE"
+
+        -- Similarly, in case of UPDATE OF, the columns can be simply parsed from the
+        -- original query. Note that UPDATE and UPDATE OF are mutually exclusive and have
+        -- the same bit set in the underlying tgtype bit field.
+        trgEvents :: Set TriggerEvent
+        trgEvents =
+          foldl (\set (mask, event) ->
```

Comment:
Please use `foldl'` here, `foldl` is useless :) https://github.com/hasura/graphql-engine/pull/2933#discussion_r328821960

Reply:
I keep forgetting about this... Will fix.